File: csharp/zugferd_facturx_xrechnung/attach_invoice_and_conv_to_zugferd/conv_to_zugferd.cs
//  conv_to_zugferd -- C# port of examples\Vb6\zugferd_facturx_xrechnung\
//    attach_invoice_and_conv_to_zugferd\conv_to_zugferd.bas
//  Converts a PDF to PDF/A-3 (FacturX Comfort), attaches the factur-x.xml
//  e-invoice and adds an output intent. Uses font-not-found and ICC-profile
//  replacement callbacks during conformance checking.
using System;
using System.IO;
using LumasPdfSdk;

class ConvToZugferd
{
    // coDefault_PDFA_3 is not exported as a const; its computed value.
    const uint coDefault_PDFA_3 = 0x50EF7F;

    static string TF = System.IO.Path.Combine(System.AppContext.BaseDirectory, "..", "..", "..", "..", "test_files");

    static int PDFError(IntPtr Data, int ErrCode, string ErrMessage, int ErrType)
    {
        Console.WriteLine(ErrMessage);
        return 0;                      // We try to continue if an error occurs
    }

    // WeightFromStyle helper: (Style and $7FF00000) shr 20 (+$800 if sign bit set)
    static int WeightFromStyle(int Style)
    {
        uint u = (uint)Style;
        int w = (int)((u & 0x7FF00000u) >> 20);
        if ((u & 0x80000000u) != 0) w += 0x800;
        return w;
    }

    static int FontNotFoundProc(IntPtr Data, IntPtr PDFFont, string FontName, int Style, int StdFontIndex, bool IsSymbolFont)
    {
        int s = Style;
        if (WeightFromStyle(s) < 500) s = (s & 0xF) | LumasPdfConsts.fsRegular;
        return LumasPdf.pdfReplaceFontW(Data, PDFFont, "Arial", s, true);
    }

    static int ReplaceICCProfileProc(IntPtr Data, TICCProfileType ProfileType, int ColorSpace)
    {
        // The most important ICC profiles are available free of charge from Adobe.
        switch (ProfileType)
        {
            case TICCProfileType.ictRGB:
                return LumasPdf.pdfReplaceICCProfileW(Data, (uint)ColorSpace, TF + "sRGB.icc");
            case TICCProfileType.ictCMYK:
                return LumasPdf.pdfReplaceICCProfileW(Data, (uint)ColorSpace, TF + "ISOcoated_v2_bas.ICC");
            default:
                return LumasPdf.pdfReplaceICCProfileW(Data, (uint)ColorSpace, TF + "gray.icc");
        }
    }

    static TErrorProc _errCb = PDFError;
    static TOnFontNotFoundProc _fontCb = FontNotFoundProc;
    static TOnReplaceICCProfile _iccCb = ReplaceICCProfileProc;

    static bool ConvertFile(IntPtr pdf, int ConvType, string InFile, string Invoice, string OutFile)
    {
        LumasPdf.pdfCreateNewPDFW(pdf, "");              // The output file is opened later
        LumasPdf.pdfSetDocInfoW(pdf, TDocumentInfo.diProducer, "");    // Keep the original producer

        uint convFlags;
        switch (ConvType)
        {
            case (int)TConformanceType.ctFacturX_Comfort:
            case (int)TConformanceType.ctFacturX_Extended:
            case (int)TConformanceType.ctFacturX_XRechnung:
                convFlags = coDefault_PDFA_3;
                break;
            default:
                return false;           // We create e-invoices in this example and nothing else.
        }

        LumasPdf.pdfCreateNewPDFW(pdf, "");              // The output file will be created later
        LumasPdf.pdfSetDocInfoW(pdf, TDocumentInfo.diProducer, "");    // Keep the original producer

        // These flags require some processing time but they are very useful.
        convFlags = LumasPdfConsts.coCheckImages | LumasPdfConsts.coRepairDamagedImages;

        // ifPrepareForPDFA is required. ifImportAsPage keeps pages from becoming templates.
        LumasPdf.pdfSetImportFlags(pdf, LumasPdfConsts.ifImportAll | LumasPdfConsts.ifImportAsPage | LumasPdfConsts.ifPrepareForPDFA);
        // if2UseProxy reduces the memory usage.
        LumasPdf.pdfSetImportFlags2(pdf, LumasPdfConsts.if2UseProxy);

        LumasPdf.pdfOpenImportFileW(pdf, InFile, (int)LumasPdfConsts.ptOpen, "");
        LumasPdf.pdfImportPDFFile(pdf, 1, 1.0, 1.0);
        LumasPdf.pdfCloseImportFile(pdf);

        // The invoice should be the first attachment. If the name is not factur-x.xml use AttachFileEx().
        int ef = LumasPdf.pdfAttachFileW(pdf, Invoice, "EN 16931 compliant invoice", false);
        if (ConvType != (int)TConformanceType.ctFacturX_XRechnung)
            LumasPdf.pdfAssociateEmbFile(pdf, TAFDestObject.adCatalog, -1, TAFRelationship.arAlternative, (uint)ef);
        else
            LumasPdf.pdfAssociateEmbFile(pdf, TAFDestObject.adCatalog, -1, TAFRelationship.arSource, (uint)ef);

        // Note that this code requires the PDF/A Extension for the classic API.
        int retval = LumasPdf.pdfCheckConformance(pdf, ConvType, convFlags, pdf, _fontCb, _iccCb);
        switch (retval)
        {
            case 1: LumasPdf.pdfAddOutputIntentW(pdf, TF + "sRGB.icc"); break;
            case 2: LumasPdf.pdfAddOutputIntentW(pdf, TF + "ISOcoated_v2_bas.ICC"); break;
            case 3: LumasPdf.pdfAddOutputIntentW(pdf, TF + "gray.icc"); break;
        }

        // No fatal error occurred?
        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            // We write the file into the application directory.
            if (!LumasPdf.pdfOpenOutputFileW(pdf, OutFile))
            {
                LumasPdf.pdfDeletePDF(pdf);
                return false;
            }
            return LumasPdf.pdfCloseFile(pdf);
        }
        return false;
    }

    static void Main()
    {
        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _errCb);

        // Non embedded CID fonts usually depend on external cmaps.
        LumasPdf.pdfSetCMapDirW(pdf, System.IO.Path.Combine(System.AppContext.BaseDirectory, "..", "..", "..", "..", "test_files") + "\\CMap", LumasPdfConsts.lcmDelayed | LumasPdfConsts.lcmRecursive);

        string outFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf");

        // The profiles Minimum, Basic, and Basic WL are not EN 16931 compliant.
        if (ConvertFile(pdf, (int)TConformanceType.ctFacturX_Comfort, TF + "test_invoice.pdf", TF + "factur-x.xml", outFile))
            Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");

        LumasPdf.pdfDeletePDF(pdf);
    }
}

File: cpp/zugferd_facturx_xrechnung/attach_invoice_and_conv_to_zugferd/conv_to_zugferd.cpp
// conv_to_zugferd -- C++ mirror of examples\Vb6\zugferd_facturx_xrechnung\
//   attach_invoice_and_conv_to_zugferd\conv_to_zugferd.bas
// Converts a PDF to PDF/A-3 (FacturX Comfort), attaches factur-x.xml and adds an
// output intent, using font-not-found and ICC-profile replacement callbacks.
#include "apputil.h"

// coDefault_PDFA_3 is not exported as a const; this is its computed value.
static const unsigned int coDefault_PDFA_3 = 0x50EF7F;

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType) {
    if (ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

// WeightFromStyle: (Style and $7FF00000) shr 20, +$800 if sign bit set.
static long WeightFromStyle(long Style) {
    long w = (long)((Style & 0x7FF00000) >> 20);
    if (Style & 0x80000000) w += 0x800;
    return w;
}

static SI32 PDF_CALL FontNotFoundProc(void* Data, void* PDFFont, const char* FontName,
                                       TFStyle Style, SI32 StdFontIndex, LBOOL IsSymbolFont) {
    long s = Style;
    if (WeightFromStyle(s) < 500) s = (s & 0xF) | (long)fsRegular;
    return pdfReplaceFontA((PPDF)Data, PDFFont, "Arial", (SI32)s, 1);
}

static SI32 PDF_CALL ReplaceICCProfileProc(void* Data, TICCProfileType ProfileType, SI32 ColorSpace) {
    switch (ProfileType) {
        case ictRGB:  return pdfReplaceICCProfileA((PPDF)Data, ColorSpace, "../../../test_files/sRGB.icc");
        case ictCMYK: return pdfReplaceICCProfileA((PPDF)Data, ColorSpace, "../../../test_files/ISOcoated_v2_bas.ICC");
        default:      return pdfReplaceICCProfileA((PPDF)Data, ColorSpace, "../../../test_files/gray.icc");
    }
}

static bool ConvertFile(PPDF pdf, int ConvType, const char* InFile, const char* Invoice, const char* OutFile) {
    unsigned int convFlags;
    pdfCreateNewPDFA(pdf, "");
    pdfSetDocInfoA(pdf, diProducer, "");

    switch (ConvType) {
        case ctFacturX_Comfort:
        case ctFacturX_Extended:
        case ctFacturX_XRechnung:
            convFlags = coDefault_PDFA_3;
            break;
        default:
            return false; // We only create e-invoices here.
    }

    pdfCreateNewPDFA(pdf, "");
    pdfSetDocInfoA(pdf, diProducer, "");

    // These flags require some processing time but are very useful.
    convFlags = coCheckImages | coRepairDamagedImages;

    pdfSetImportFlags(pdf, ifImportAll | ifImportAsPage | ifPrepareForPDFA);
    pdfSetImportFlags2(pdf, if2UseProxy);

    pdfOpenImportFileA(pdf, InFile, ptOpen, "");
    pdfImportPDFFile(pdf, 1, 1, 1);
    pdfCloseImportFile(pdf);

    int ef = pdfAttachFileA(pdf, Invoice, "EN 16931 compliant invoice", 0);
    if (ConvType != ctFacturX_XRechnung)
        pdfAssociateEmbFile(pdf, adCatalog, -1, arAlternative, ef);
    else
        pdfAssociateEmbFile(pdf, adCatalog, -1, arSource, ef);

    int retval = pdfCheckConformance(pdf, ConvType, convFlags, pdf, FontNotFoundProc, ReplaceICCProfileProc);
    switch (retval) {
        case 1: pdfAddOutputIntentA(pdf, "../../../test_files/sRGB.icc"); break;
        case 2: pdfAddOutputIntentA(pdf, "../../../test_files/ISOcoated_v2_bas.ICC"); break;
        case 3: pdfAddOutputIntentA(pdf, "../../../test_files/gray.icc"); break;
    }

    if (pdfHaveOpenDoc(pdf) != 0) {
        if (pdfOpenOutputFileA(pdf, OutFile) == 0) { pdfDeletePDF(pdf); return false; }
        return pdfCloseFile(pdf) != 0;
    }
    return false;
}

int main() {
    ChdirToExe();
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, PDFError);

    // Non-embedded CID fonts usually depend on external cmaps.
    pdfSetCMapDirA(pdf, "../../../Resource/CMap", lcmDelayed | lcmRecursive);

    const char* outFile = "out.pdf";
    if (ConvertFile(pdf, ctFacturX_Comfort, "../../../test_files/test_invoice.pdf",
                    "../../../test_files/factur-x.xml", outFile))
        printf("PDF file \"%s\" successfully created!\n", outFile);

    pdfDeletePDF(pdf);
    return 0;
}
